doc/aaa/ldap.md

```diff
@@ -0,0 +1,2617 @@
+                                         | LDAP Name Service,   and   |
+                                     +-->| LDAP sudoers Authorization |
+                                     |   |          Servers           |
+                                     |   |                            |
+   +------------+    +-----------+   |   +----------------------------+
+   |            |    |           |   |
+   |            |    |           |   |   +----------------------------+
+   |            |    |           |   |   |                            |
+   | SSH Client |<-->| SONiC DUT |<--+-->| TACACS+   Authentication   |
+   |            |    |           |   |   |         Servers            |
+   |            |    |           |   |   +----------------------------+
+   |            |    |           |   |
+   |            |    |           |   |
+   +------------+    +-----------+   |   +----------------------------+
+                                     |   |                            |
+                                     +-->| LDAP Login Authorization   |
+                                         |         Servers            |
+                                         +----------------------------+
+
+```
+
+```
+
+
+; LDAP Name Service parameters
+sonic(config)# ldap-server base "dc=example,dc=com"
+sonic(config)# aaa name-service passwd ldap
+sonic(config)# aaa name-service shadow ldap
+sonic(config)# aaa name-service group ldap
+
+; Use this server for name-service and sudo-ldap authorization only, with one
+; reconnect attempt
+sonic(config)# ldap-server rsa.loc.example.net use-type nss,sudo ldap-retry 1
+
+
+; Logon Authorization from LDAP
+; Use this server for login authorization only
+sonic(config)# ldap-server 10.59.1.9 use-type pam
+
+sonic(config)# ldap-server pam-groupdn "cn=t1.loc.example.com,ou=hostaccess,dc=example,dc=com"
+sonic(config)# ldap-server pam-member-attribute "uniqueMember"
+sonic(config)# aaa authorization login ldap
+
+; sudo-ldap authorization
+sonic(config)# ldap-server sudoers-base "dc=example,dc=com"
+sonic(config)# aaa name-service sudoers ldap
+
+; Authentication from TACACS+
+sonic(config)# tacacs-server 10.59.1.15
+sonic(config)# aaa authentication failthrough enable
+sonic(config)# aaa authentication login-method tacacs+ local
+
+```
+
+# Flow Diagrams
+
+Please see the diagrams in section [Design](#design)
+
+# Error Handling
+
+PAM and NSS modules return errors as per [Linux-PAM](#Linux-PAM), and
+[NSS](#NSS-Reference) respectively.
+
+# Serviceability and Debug
+
+The LDAP PAM and NSS modules can be debugged using the "debug" option
+of AAA. (i.e. config aaa debug enable)
+
+# Warm Boot Support
+
+N/A
+
+# Scalability
+
+A maximum of 8 LDAP servers can be configured.
+
+# Unit Test
+
+## Prerequisites
+
+```
+  LDAP Server with User, Group information.
+  TACACS+/RADIUS Server with User passwords configure.
+```
+
+### LDAP data
+
+For sudo authorization (*sudoRole* objectclass), the LDAP server schema
```

Comment:
It would be interesting to see how Cisco defines the roles on an LDAP server. They must have defined a schema for that. Since customers would already have configured their LDAP servers to operate with Cisco equipment we really need to follow the same schema as Cisco.

Reply:
I believe Industry Standard uses an attribute in the user's entry in the LDAP directory to obtain role information. 

The current proposal is to use the group membership which is the same way that local users' roles are determined (RBAC Phase II document -- "supplementary groups are used to allow access to additional functionalities"). So this is in keeping with the general account management design. 

Another approach could be to have a role attribute in the user's LDAP entry. 

(Both these are mentioned in the Session Setup section of the document. The first one is being used in this release. For following releases, the second approach could be fitted in using existing LDAP maps (or defining new ones by extending the set of LDAP maps).

Eg: **LDAP_MAP** | **attribute**|role  --> { '**to**' : ' groups' }